buffer/src/main/java/io/netty/buffer/api/ByteCursor.java

```diff
@@ -0,0 +1,108 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.buffer.api;
+
+import io.netty.util.ByteProcessor;
+
+/**
+ * The ByteCursor scans through a sequence of bytes.
+ * This is similar to {@link ByteProcessor}, but for external iteration rather than internal iteration.
+ * The external iteration allows the callers to control the pace of the iteration.
+ * The API includes methods for reading {@code long}s as a batch of 8 bytes.
+ * The long values are always in big-endian format, so that the highest-order byte in the long value, contain the byte
+ * that would otherwise have been returned by the next call to {@link #getByte()}.
+ */
+public interface ByteCursor {
+    /**
+     * Check if the iterator has at least 8 bytes left, and if so, read those 8 bytes and move the cursor forward.
+     * The bytes are packed as a {@code long} value in big-endian format, such that the highest-order byte
+     * in the long, is the byte that would otherwise have been returned by the next call to {@link #getByte()},
+     * after a call to {@link #readByte()}.
+     * The bytes (as a {@code long}) will then be available through the {@link #getLong()} method.
+     * <p>
+     * Note that when this method returns {@code false}, the {@link #readByte()} can still return {@code true}.
+     * It is recommended to have any long-processing loop be followed by a byte-processing loop for the 7 or fewer
+     * bytes that might form a tail in the cursor.
+     * <p>
+     * Also note that this method will not influence what is returned the {@link #getByte()} method.
+     *
+     * @return {@code true} if the cursor read 8 bytes and moved forward, otherwise {@code false}.
+     */
+    boolean readLong();
+
+    /**
+     * Return the last 8 bytes read by {@link #readLong()}.
+     * If {@link #readLong()} has not been called on this cursor before, then {@code -1} is returned.
+     * <p>
+     * The long value is in the big-endian format, such that the highest-order by of the long value, is the byte that
+     * would otherwise have been produced by a {@link #readByte()} / {@link #getByte()} pair.
+     * This means that cursors that iterate in reverse, e.g. from {@link Buffer#openReverseCursor()}, return longs in a
+     * "mirrored" or "reversed" big-endian format.
+     *
+     * @return The 8 bytes, in big-endian format, that was read by the most recent successful call to
+     * {@link #readLong()}.
+     */
+    long getLong();
+
+    /**
+     * Check if the iterator has at least one byte left, and if so, read that byte and move the cursor forward.
+     * The byte will then be available through the {@link #getByte()}.
+     * <p>
+     * Note that this method will not influence what is returned from the {@link #getLong()} method.
+     *
+     * @return {@code true} if the cursor read a byte and moved forward, otherwise {@code false}.
+     */
+    boolean readByte();
```

Comment:
hasNext() ?

Reply:
I've intentionally not reused the iterator names in order to highlight the difference in behaviour between iterators and cursors.

Reply:
Maybe just `next` then?   java.sql.ResultSet is a cursor and uses the "next" idiom.    I haven't seen that many uses of cursors outside of DBs, and they don't typically overload the name like this.   I think this might be surprising to more people than just me.

Reply:
@normanmaurer What do you think? We previously had a `next*` naming convention for these methods.